unison-share-api/src/Unison/Server/CodebaseServer.hs

```diff
@@ -146,16 +147,72 @@ data BaseUrl = BaseUrl
     urlPort :: Port
   }
 
-data BaseUrlPath = UI | Api
+data NameOrHashRef = Name [Text] | Hash Text
+
+data DefinitionReference
+  = TermReference NameOrHashRef
+  | TypeReference NameOrHashRef
+  | AbilityConstructorReference NameOrHashRef
+  | DataConstructorReference NameOrHashRef
+
+data BaseUrlPath
+  = UI
+      { namespace :: Maybe [Text],
+        definition :: Maybe DefinitionReference
+      }
+  | Api
 
 instance Show BaseUrl where
   show url = urlHost url <> ":" <> show (urlPort url) <> "/" <> (URI.encode . unpack . urlToken $ url)
 
 urlFor :: BaseUrlPath -> BaseUrl -> String
```

Comment:
Would you mind adding some doc-tests here? I find they really help with readability for expecting what the output looks like in the future and are a really cheap form of test to write for pure functions 😄